Document the resource quota working and steps require to enforce  resource quota on ephemeral storage (#35011)
kubernetes/website: content/en/docs/tasks/administer-cluster/manage-resources/quota-memory-cpu-namespace.md

pods "quota-mem-cpu-demo-2" is forbidden: exceeded quota: mem-cpu-demo,
requested: requests.memory=700Mi,used: requests.memory=600Mi, limited: requests.memory=1Gi
```
{{< warning >}}

> User needs to specify limit in pod spec. If the user doesn't specify limit in pod spec, then resourcequota is not enforced on that resource. So for enforcement of resourcequota on resources, user must specify limit in pod.

Reviewer: Thanks for documenting it which is helpful in understanding the quota current behaviour. Some suggestion in the sentence:

{{< warning >}}

To make the resource quota work, two things need to be done.
1. admin set the resource quota limit
2. User needs to mention the resource quota limit in pod spec.

If the user doesn't specify the resource limit in pod spec, then the resource quota is not enforced on that resource.
For example, if the admin set the ephemeral-storage quota and if the user doesn't specify the ephemeral-storage limit in the pod creation spec then the ephemeral-storage resource quota will not be enforced. Thus the user can use n number of ephemeral storage.

{{< /warning >}}

Author: @gmannos Thank you for the suggestions i have updated the title and comments as you suggested .PTAL
Thanks!

Reviewer 2: > User needs to mention the resource quota limit in pod spec.

If an admission plugin sets default limits for resources, then that defaulting is also adequate. See https://kubernetes.io/docs/concepts/policy/limit-range/ for an example.

pods "quota-mem-cpu-demo-2" is forbidden: exceeded quota: mem-cpu-demo,
requested: requests.memory=700Mi,used: requests.memory=600Mi, limited: requests.memory=1Gi
```

> {{< warning >}}

Reviewer 2: If this _is_ correct, use a `note` not a `warning`.

Author: @sftim Sure , addressed this change in latest commit.

To make the resource quota work, two things need to be done.

1.admin set the resource quota limit
2.User needs to mention the resource quota limit in pod spec.

> If the user doesn't specify the resource limit in pod spec, then the resource quota is not enforced on that resource. For example, if the admin set the ephemeral-storage quota and if the user doesn't specify the ephemeral-storage limit in the pod creation spec then the ephemeral-storage resource quota will not be enforced. Thus the user can use n number of ephemeral storage.

Reviewer 3: I'm not sure this is technically correct.
Each type/kind of resources are treated in different ways by different controllers. Maybe ephemeral storage is an exception we need to call out, but I don't think resource quota limit need to be mentioned in to Pod `spec` explicitly for all resource types.

Author: Hi @tengqm  yes , i have checked with other resources also like secrets and found that resource quota enforce on secrets as 
`error: failed to create secret secrets "db-user" is forbidden: exceeded quota: object-counts, requested: secrets=1, used: secrets=2, limited: secrets=2` . It seems like this is an exception for ephemeral-storage and i have added corresponding changes . Please take a look .
Thanks!!

Author: I have modified the PR title for enforcing resource quota limits on ephemeral storage.

pods "quota-mem-cpu-demo-2" is forbidden: exceeded quota: mem-cpu-demo,
requested: requests.memory=700Mi,used: requests.memory=600Mi, limited: requests.memory=1Gi
```
{{< note >}}

> This is an exception for enforcing resourcequota on ephemeral-storage.

Reviewer: space please

> This is an exception for enforcing resourcequota on ephemeral-storage.

Reviewer: space in resourcequota

Author: @gmannos Done !

Author: @gmannos I have modified the content in the note as suggested in [comment](https://github.com/kubernetes/website/pull/35011#discussion_r940922995) for wrapping it up . Please take a look again.

Thanks!

Reviewer 2: When reviewing, aim to follow our [style guide](https://kubernetes.io/docs/contribute/style/style-guide/), which I think would recommend.

```markdown
There is an exception for enforcing ResourceQuota on ephemeral storage.
```

> This is an exception for enforcing resource quota on ephemeral-storage.

Reviewer 2: Is this an exception, or are we describing Kubernetes normal mechanism for resource requests and limits?

Author: @sftim  I have checked with other resources i.e CPU and Memory , we explicitly need to mention these values in pod spec, if we don't it throw as `Error creating: pods "ephpod-668668fc76-4rm6p" is forbidden: failed quota: ephtest: must specify limits.cpu` . But When i am trying with ephemeral-storage and don't provide request/limit in pod spec then pod is created but resource quota not enforced i.e i can use any number of ephemeral storage . So unless we provided request/limit in pod spec , resource quota is not enforced on ephemeral storage.

Reviewer 3: This is -> There is ?

Reviewer 2: > I have checked with other resources i.e CPU and Memory , we explicitly need to mention these values in pod spec, if we don't it throw as `Error creating: pods "ephpod-668668fc76-4rm6p" is forbidden: failed quota: ephtest: must specify limits.cpu` . But When i am trying with ephemeral-storage and don't provide request/limit in pod spec then pod is created but resource quota not enforced i.e i can use any number of ephemeral storage . So unless we provided request/limit in pod spec , resource quota is not enforced on ephemeral storage.

That sounds like there might be a LimitRange or other restriction. As far as I know a ResourceQuota _on its own_ doesn't force you to specify CPU or memory resources (I might be wrong and haven't checked).

To make the resource quota work on ephemeral-storage, two things need to be done.

> 1.admin set the ephemeral-storage resource quota limit

Reviewer 3: ```suggestion
1. an admin sets the resource quota for ephemeral-storage
```

> 2.User needs to mention the ephemeral-storage resource quota limit in pod spec.

Reviewer 3: ```suggestion
1. User needs to mention the ephemeral-storage resource limit in the Pod spec.
```

> * User needs to mention the ephemeral-storage resource limit in the Pod spec.

Reviewer 2: ```suggestion
* You also need to define a limit for `ephemeral-storage` in the Pod spec.
  A cluster administrator can define a [LimitRange](/docs/concepts/policy/limit-range/)
  to enforce that you do define that limit.
```

Author: Done

> 2. User needs to mention the ephemeral-storage resource limit in the Pod spec.

Reviewer 3: Do NOT use explicit numbers for ordered list, unless you have a strong reason to do so.

```suggestion
1. User needs to mention the ephemeral-storage resource limit in the Pod spec.
```

Author: Done

> If the user doesn't specify the ephemeral-storage resource limit in pod spec, then the resource quota is not enforced on ephemeral-storage. Thus the user can use n number of ephemeral storage.

Reviewer 3: ```suggestion
If the user doesn't specify the ephemeral-storage resource limit in the Pod spec,
the resource quota is not enforced on ephemeral-storage. Thus the user can use
any amount of ephemeral-storage.
```

Author: @tengqm i have addressed the suggestions . PTAL

> If the user doesn't specify the ephemeral-storage resource limit in the Pod spec, the resource quota is not enforced on ephemeral-storage. Thus the user can use any amount of ephemeral-storage.

Reviewer 3: please DO wrap the long lines like this. git tracks changes by lines.

```suggestion
If the user doesn't specify the ephemeral-storage resource limit in the Pod spec,
the resource quota is not enforced on ephemeral-storage. Thus the user can use
any amount of ephemeral-storage.
```

Reviewer 3: This is not resolved.
/hold

> If the user doesn't specify the ephemeral-storage resource limit in pod spec, then the resource quota is not enforced on ephemeral-storage. For example, if the admin set the ephemeral-storage quota and if the user doesn't specify the ephemeral-storage limit in the pod creation spec then the  resource quota on ephemeral-storage will not be enforced. Thus the user can use n number of ephemeral storage.

Reviewer 3: Please wrap the long line; please fix grammar of the last sentence.

Author: @tengqm i gave a try to wrap it up. Please take a look and let me know if this sounds ok to you.
Thanks!!

pods "quota-mem-cpu-demo-2" is forbidden: exceeded quota: mem-cpu-demo,
requested: requests.memory=700Mi,used: requests.memory=600Mi, limited: requests.memory=1Gi
```
{{< note >}}
This is an exception for enforcing resource quota on ephemeral-storage.

To make the resource quota work on ephemeral-storage, two things need to be done.

* an admin sets the resource quota for ephemeral-storage
* You also need to define a limit for `ephemeral-storage` in the Pod spec.
  A cluster administrator can define a [LimitRange](/docs/concepts/policy/limit-range/)
  to enforce that you do define that limit.

> If the user doesn't specify the ephemeral-storage resource limit in the Pod spec, the resource quota is not enforced on ephemeral-storage. Thus the user can use any amount of ephemeral-storage.

Reviewer 3: Please MANUALLY WRAP THE LONG LINE.

Author: @tengqm Done.
